packages/grid/x-data-grid-generator/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,271 @@
+    firstRowIndex = Math.max(firstRowIndex, 0); // if cursor not found return 0
+    lastRowIndex = firstRowIndex + pageSize;
+
+    nextCursor = lastRowIndex >= filteredRows.length ? undefined : filteredRows[lastRowIndex].id;
+  } else {
+    firstRowIndex = page * pageSize;
+    lastRowIndex = (page + 1) * pageSize;
+  }
+  const response: FakeServerResponse = {
+    returnedRows: filteredRows.slice(firstRowIndex, lastRowIndex),
+    nextCursor,
+    rowNumber,
+  };
+
+  return new Promise<FakeServerResponse>((resolve) => {
+    setTimeout(() => {
+      resolve(response);
+    }, delay); // simulate network latency
+  });
+};
+
+interface FakeServerResponse {
+  returnedRows: GridRowModel[];
+  nextCursor?: string;
+  rowNumber: number;
+}
+
+interface PageInfo {
+  totalRowCount?: number;
+  nextCursor?: string;
+  pageSize?: number;
+}
+
+export interface ServerOptions {
+  minDelay: number;
+  maxDelay: number;
+  useCursorPagination?: boolean;
+}
+
+export interface QueryOptions {
+  cursor?: GridRowId;
+  page?: number;
+  pageSize?: number;
+  // TODO: implement the behavior liked to following models
+  filterModel?: GridFilterModel;
+  sortModel?: GridSortModel;
+}
+
+const DEFAULT_DATASET_OPTIONS: UseDemoDataOptions = {
+  dataSet: 'Commodity',
+  rowLength: 100,
+  maxColumns: 6,
+};
+
+const DEFAULT_SERVER_OPTIONS: ServerOptions = {
+  minDelay: 100,
+  maxDelay: 300,
+  useCursorPagination: true,
+};
+
+export const createFakeServer = (
+  dataSetOptions?: Partial<UseDemoDataOptions>,
+  serverOptions?: Partial<ServerOptions>,
+) => {
+  const dataSetOptionsWithDefault = { ...DEFAULT_DATASET_OPTIONS, ...dataSetOptions };
+  const serverOptionsWithDefault = { ...DEFAULT_SERVER_OPTIONS, ...serverOptions };
+
+  const columns = getColumnsFromOptions(dataSetOptionsWithDefault);
+  const initialState = getInitialState(dataSetOptionsWithDefault, columns);
+
+  const defaultColDef = getGridDefaultColumnTypes();
+  const columnsWithDefaultColDef = columns.map((column) => ({
+    ...defaultColDef[column.type || 'string'],
+    ...column,
+  }));
+
+  const useQuery = (queryOptions: QueryOptions) => {
+    const {
+      data: { rows },
+      loading: dataGenerationIsLoading,
+    } = useDemoData(dataSetOptionsWithDefault);
+
+    const queryOptionsRef = React.useRef(queryOptions);
+    const [response, setResponse] = React.useState<{
+      pageInfo: PageInfo;
+      data: GridRowModel[];
+    }>({ pageInfo: {}, data: [] });
+    const [isLoading, setIsLoading] = React.useState<boolean>(false);
+
+    React.useEffect(() => {
+      if (dataGenerationIsLoading) {
+        // dataset is not ready
+        return () => {};
+      }
+
+      queryOptionsRef.current = queryOptions;
+      let active = true;
+
+      setIsLoading(true);
+      setResponse({ pageInfo: {}, data: [] });
+      loadServerRows(rows, queryOptions, serverOptionsWithDefault, columnsWithDefaultColDef).then(
```

Comment:
Can we convert it to `async / await`?

Reply:
Would it be easier to understand? I personally prefer to read all the code in its running order, instead of defining an async function of two lines that will be called at the end

Reply:
I guess it is more of personal preferences. I think In the codebase we opt-in for using async / await. Also, you can make a self-executing function in the `useEfect` and not have it being defined and called immediately after.

Reply:
done